sim/state.ts

```diff
@@ -0,0 +1,340 @@
+	serializeBattle(battle: Battle): /* Battle */ AnyObject {
+		const state: /* Battle */ AnyObject = this.serialize(battle, BATTLE, battle);
+		state.field = this.serializeField(battle.field);
+		state.sides = new Array(battle.sides.length);
+		for (const [i, side] of battle.sides.entries()) {
+			state.sides[i] = this.serializeSide(side);
+		}
+		state.prng = battle.prng.seed;
+		state.hints = Array.from(battle.hints);
+		return state;
+	}
+
+	// Deserialization can only really be done on the root Battle object as
+	// the leaf nodes like Side or Pokemon contain backreferences to Battle
+	// but don't contain the information to fill it in because the cycles in
+	// the graph have been serialized as references. Once deserialzized, the
+	// Battle can than be restarted (and provided with a `send` function for
+	// receiving updates).
+	deserializeBattle(serialized: string | /* Battle */ AnyObject): Battle {
+		const state: /* Battle */ AnyObject =
+			typeof serialized === 'string' ? JSON.parse(serialized) : serialized;
+		const options = {
+			formatid: state.format,
+			seed: state.prngSeed,
+			rated: state.rated,
+			debug: state.debugMode,
+			// We need to tell that Battle we're creating that it's been
+			// deserialized so that it allows us to populate it correctly and
+			// doesn't attempt to start playing out until we're ready.
+			deserialized: true,
+			strictChoices: state.strictChoices,
+		};
+		for (const side of state.sides) {
+			const team = side.team.split(side.team.length > 9 ? ',' : '');
+			// @ts-ignore - index signature
+			options[side.id] = {
+				name: side.name,
+				avatar: side.avatar,
+				team: team.map((p: string) => side.pokemon[Number(p) - 1].set),
+			};
+		}
+		// We create the Battle, allowing it to instantiate the Field/Side/Pokemon
+		// objects for us. The objects it creates will be incorrect, but we descend
+		// down through the fields and repopulate all of the objects with the
+		// correct state.
+		const battle = new Battle(options);
+		this.deserialize(state, battle, BATTLE, battle);
+		this.deserializeField(state.field, battle.field);
+		for (const [i, side] of state.sides.entries()) {
+			this.deserializeSide(side, battle.sides[i]);
+		}
+		battle.prng = new PRNG(state.prng);
+		// @ts-ignore - readonly
+		battle.hints = new Set(battle.hints);
+		return battle;
+	}
+
+	serializeField(field: Field): /* Field */ AnyObject {
+		return this.serialize(field, FIELD, field.battle);
+	}
+
+	private deserializeField(state: /* Field */ AnyObject, field: Field) {
+		this.deserialize(state, field, FIELD, field.battle);
+	}
+
+	serializeSide(side: Side): /* Side */ AnyObject {
+		const state: /* Side */ AnyObject = this.serialize(side, SIDE, side.battle);
+		state.pokemon = new Array(side.pokemon.length);
+		const team = new Array(side.pokemon.length);
+		for (const [i, pokemon] of side.pokemon.entries()) {
+			state.pokemon[i] = this.serializePokemon(pokemon);
+			team[side.team.indexOf(pokemon.set)] = i + 1;
```

Comment:
I would suggest to switch these to use ``map()``. I find this code pretty confusing.

It's equivalent to ``const team = side.team.map(set => pokemon.set.indexOf(set) + 1)``... ... Is it???? :[